File: bindings/rust/opentlv/examples/write.rs
//! Builds the same nested BER-TLV document `parse.rs` reads, encoding the
//! innermost elements first and using each encoded result as the next
//! level's value: the standard way to build constructed TLV bottom-up.
//!
//! Run with `cargo run --example write` from `bindings/rust`.

use opentlv::{Format, Result, Tag, Writer};

/// Same bytes as `parse.rs`'s document.
const EXPECTED: [u8; 12] = [
    0x6F, 0x0A, 0x84, 0x03, 0x41, 0x42, 0x43, 0xA5, 0x03, 0x50, 0x01, 0x01,
];

fn main() -> Result<()> {
    let mut df_name_buf = [0u8; 5];
    let mut df_name = Writer::with_format(&mut df_name_buf, Format::Ber);
    df_name.write(&Tag::from_bytes(&[0x84]), b"ABC")?;

    let mut label_buf = [0u8; 3];
    let mut label = Writer::with_format(&mut label_buf, Format::Ber);
    label.write(&Tag::from_bytes(&[0x50]), &[0x01])?;

    let mut proprietary_buf = [0u8; 8];
    let mut proprietary = Writer::with_format(&mut proprietary_buf, Format::Ber);
    proprietary.write(&Tag::from_bytes(&[0xA5]), label.written())?;

    let mut value = Vec::new();
    value.extend_from_slice(df_name.written());
    value.extend_from_slice(proprietary.written());

    let mut document_buf = [0u8; 20];
    let mut document = Writer::with_format(&mut document_buf, Format::Ber);
    document.write(&Tag::from_bytes(&[0x6F]), &value)?;

    println!(
        "wrote {} bytes: {:02X?}",
        document.written().len(),
        document.written()
    );
    assert_eq!(document.written(), &EXPECTED);
    Ok(())
}

File: bindings/python/opentlv/examples/write.py
"""Builds the same nested BER-TLV document parse.py reads, encoding the
innermost elements first and using each encoded result as the next level's
value: the standard way to build constructed TLV bottom-up.

Run with `python examples/write.py` from `bindings/python/opentlv`, after
installing both packages.
"""

import opentlv

# Same bytes as parse.py's document.
EXPECTED = bytes([0x6F, 0x0A, 0x84, 0x03, 0x41, 0x42, 0x43, 0xA5, 0x03, 0x50, 0x01, 0x01])


def main() -> None:
    df_name = opentlv.Writer(opentlv.Format.BER)
    df_name.write(opentlv.Tag(b"\x84"), b"ABC")

    label = opentlv.Writer(opentlv.Format.BER)
    label.write(opentlv.Tag(b"\x50"), b"\x01")

    proprietary = opentlv.Writer(opentlv.Format.BER)
    proprietary.write(opentlv.Tag(b"\xa5"), label.bytes())

    document = opentlv.Writer(opentlv.Format.BER)
    document.write(opentlv.Tag(b"\x6f"), df_name.bytes() + proprietary.bytes())

    encoded = document.bytes()
    print(f"wrote {len(encoded)} bytes: {encoded.hex(' ').upper()}")
    assert encoded == EXPECTED


if __name__ == "__main__":
    main()
